### ISSUE TITLE
Swarm has deadlock with the scheduler

### ISSUE THREAD

**Issue by aanm** [docker/swarm#1194]
I think swarm has a deadlock with the scheduler if a request is lost between swarm master and the docker daemon.
```
$ DOCKER_HOST=localhost:2375 docker info
Containers: 31
Images: 60
Role: primary
Strategy: spread
Filters: affinity, health, constraint, port, dependency
Nodes: 3
 localhost.localdomain: 192.168.50.1:2371
  └ Containers: 11
  └ Reserved CPUs: 0 / 8
  └ Reserved Memory: 0 B / 16.23 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=4.1.6-200.fc22.x86_64, operatingsystem=Fedora 22 (Twenty Two), storagedriver=overlay
 node1: 192.168.50.5:2371
  └ Containers: 10
  └ Reserved CPUs: 0 / 2
  └ Reserved Memory: 0 B / 1.019 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.13.0-24-generic, operatingsystem=Ubuntu 14.04 LTS, storagedriver=aufs
 node2: 192.168.50.6:2371
  └ Containers: 10
  └ Reserved CPUs: 0 / 2
  └ Reserved Memory: 0 B / 1.019 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.13.0-24-generic, operatingsystem=Ubuntu 14.04 LTS, storagedriver=aufs
CPUs: 12
Total Memory: 18.27 GiB
Name: 38dc57b17c85

$ DOCKER_HOST=localhost:2375 docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:35:49 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      swarm/0.4.0
 API version:  1.16
 Go version:   go1.4.2
 Git commit:   d647d82
 Built:        
 OS/Arch:      linux/amd64
```

https://gist.github.com/aanm/cd073fe77f74d63732e4

ping @abronan 

EDIT: This occurs with replication mode

**Comment by aanm** [comment]
Probably related with https://github.com/docker/swarm/issues/786 but I was able to do `docker info`, `docker ps -a` etc. What I wasn't able to do was `docker create` or `docker run`

**Comment by aanm** [comment]
Turns out it happens to without replication mode

**Comment by abronan** [comment]
I'm not yet 100% sure but this might not be *only* a Swarm issue, docker `1.8.2-rc1` fixed a few issues related to `docker pull`:

- Fix ^C on docker pull (*on the cli side*)
- Fix docker pull issues on client disconnection.

So this could also be an intermittent (and rare) issue on pull with `docker run` (which goes through the scheduling process), the pull stays blocked and the scheduler remains also locked because of this. The client is indefinitely waiting for the pull to finish. The scheduler being locked is a huge issue in that case because it prevents all the other requests from being processed (on your stacktrace we can see all the `CreateContainer` requests waiting for the `Lock` on the scheduler). And adding a timeout would not solve the issue because this might just be a very slow pull.

This also explains why you can still do a `docker info` or `docker ps -a` but not a `docker create` or `docker run`. Both `docker create` and `docker run` needs the scheduler which in this case remains locked by some other request.

As of now I'm testing swarm against the latest docker `1.8.2-rc1` and I can't reproduce. So this might confirm this theory..

The issue is also on the scheduler which is being locked down to avoid races on resource reservation but updating to the latest docker `rc` version *might* do the trick for now while we think about a better way to handle those issues on the scheduler.

If you can test against the latest `rc` and confirm that would be great @aanm :)

/cc @vieux @aluzzardi

**Comment by aanm** [comment]
@abronan I'll test it later, btw I can confirm that there isn't a single pull being done by the requests since all images are already installed.
My setup is something like this:
```
docker run hello-world 
    |
    v
docker swarm master (primary)
    |
    v
remote private server that is listening for requests 
    |
    v
Connection lost so there isn't a reply and swarm master keeps waiting for reply so it might be locked?

```

**Comment by abronan** [comment]
@aanm thanks for checking. Yes the pull issue I mentioned has virtually the same effect than dropping a connection in the middle of a create request: The scheduler could stay locked. This is a general issue with the scheduler if you are dropping connection between the Manager and the Agents. Was trying to pinpoint what could possibly trigger this situation without dropping the connection. But those locks on create and run should be removed altogether.

**Comment by abronan** [comment]
This should be fixed by #1261, Closing this one.

@aanm Thanks! Let us know if you have any more issue with master :)
